tests.py

```diff
@@ -1,15 +1,800 @@
+                'exception': None
+            },
+            {
+                'name': 'test_above_api_error_threshold',
+                'api': CommCareAPIHandler(
+                    'test_domain',
+                    'test_domain-api-key',
+                    datetime.strptime(
+                        '2024-01-01 00:00:00', '%Y-%m-%d %H:%M:%S')
+                ),
+                'parameters': {
+                    'method': raise_api_error
+                },
+                'repeat': 3,
+                'expect_exception': True,
+                'exception': util.APIError("Request failed for data type 2!", 400)
+            }
+        ]
+
+        for test_case in test_data:
+            with self.subTest(test_case['name']):
+                print(f'  Running subtest {test_case["name"]}')
+                try:
+                    for i in range(test_case['repeat']):
+                        test_case['api']._perform_method(
+                            test_case['parameters']['method'],
+                            i
+                        )
+                    self.assertFalse(test_case['expect_exception'])
+                except Exception as e:
+                    self.assertTrue(test_case['expect_exception'])
+                    self.assertEqual(str(test_case['exception']), str(e))
+
+
+# TODO: Implement CommCareAPIHandlerPull tests
+"""
+class TestCommCareAPIHandlerPull(unittest.TestCase):
+    def test_commcareapihandlerpull_filepath(self):
         self.assertEqual(True, False)
 
-class TestCommCareAPIHandlerPush(unittest.TestCase):
-    def test_basic(self):
+    def test_commcareapihandlerpull_get_last_job_success_time(self):
         self.assertEqual(True, False)
 
+    def test_commcareapihandlerpull_last_job_success_time_filepath(self):
+        self.assertEqual(True, False)
+
+    def test_commcareapihandlerpull_get_date_range(self):
+        self.assertEqual(True, False)
+
+    def test_commcareapihandlerpull_get_initial_parameters_for_data_type(self):
+        self.assertEqual(True, False)
+
+    def test_commcareapihandlerpull_store_in_s3(self):
+        self.assertEqual(True, False)
+
+    def test_commcareapihandlerpull_pull_data(self):
+        self.assertEqual(True, False)
+
+    def test_commcareapihandlerpull_save_run_time(self):
+        self.assertEqual(True, False)
+
+    def test_commcareapihandlerpull_pull_data_for_domain(self):
+        self.assertEqual(True, False)
+"""
+
+
+class TestCommCareAPIHandlerPush(unittest.TestCase):
+    def test_commcareapihandlerpush_filepath(self):
+        print('*** Running test_commcareapihandlerpush_filepath ***')
+        test_data = [
+            {
+                'name': 'test_path',
+                'api': CommCareAPIHandlerPush(
+                    'test_domain',
+                    'test_domain-api-key',
+                    datetime.strptime(
+                        '2024-01-01 00:00:00', '%Y-%m-%d %H:%M:%S')
+                ),
```

Comment:
I can do that change - it should be quick and easy!  I made it a property of the test to prevent persistent information from being saved to the api handler between different tests by making each use one freshly made from scratch, but for these tests that isn't really relevant, so it makes more sense to make it a property of the class like you said, or at least only implement it once in each test in the class rather than implementing a new one for each subtest as well.